concurrent_test.go

```diff
@@ -177,41 +199,53 @@ type readWorker struct {
 
 	minReadInterval time.Duration
 	maxReadInterval time.Duration
-	errCh           chan error
-	stopCh          chan struct{}
+
+	recordingCh chan historyRecord
+	errCh       chan error
+	stopCh      chan struct{}
 
 	t *testing.T
 }
 
-func (reader *readWorker) run() error {
+func (r *readWorker) run() error {
 	for {
 		select {
-		case <-reader.stopCh:
-			reader.t.Log("Reading transaction finished.")
+		case <-r.stopCh:
+			r.t.Log("Reading transaction finished.")
 			return nil
 		default:
 		}
 
-		err := reader.db.View(func(tx *bolt.Tx) error {
-			b := tx.Bucket(reader.bucket)
+		err := r.db.View(func(tx *bolt.Tx) error {
+			b := tx.Bucket(r.bucket)
 
-			selectedKey := reader.keys[mrand.Intn(len(reader.keys))]
+			selectedKey := r.keys[mrand.Intn(len(r.keys))]
 			initialVal := b.Get([]byte(selectedKey))
-			time.Sleep(randomDurationInRange(reader.minReadInterval, reader.maxReadInterval))
+			time.Sleep(randomDurationInRange(r.minReadInterval, r.maxReadInterval))
 			val := b.Get([]byte(selectedKey))
 
 			if !reflect.DeepEqual(initialVal, val) {
 				return fmt.Errorf("read different values for the same key (%q), value1: %q, value2: %q",
 					selectedKey, formatBytes(initialVal), formatBytes(val))
 			}
 
+			clonedVal := make([]byte, len(val))
+			copy(clonedVal, val)
+
+			r.recordingCh <- historyRecord{
```

Comment:
In the robustness tests I wanted to avoid synchronizing on every operation. Reason is that slow recording can result in one fast worker starving other ones. For example reads being faster then writes result in writes being starved. 

I would recommend collecting local history in every worker separately and synchronizing only after all work being done.

Reply:
Currently there is no starving issue, because each worker will sleep for a while between two transactions. But I agree we need improve it because I will support no sleep worker. Let me improve in followup PR.

Reply:
Done.